prometheus/prometheus.yml

```diff
@@ -0,0 +1,47 @@
+# Prometheus global config
+global:
+  scrape_interval: 30s # Scrape services for updated metrics every 30s. Default is 1m.
+  evaluation_interval: 30s # Evaluate rules every 30s. Default is 1m.
+  # scrape_timeout is set to the global default (10s).
+
+# Alertmanager configuration
+alerting:
+  alertmanagers:
+    - static_configs:
+        - targets:
+          # deploy-sourcegraph-docker does not yet use prometheus alerts
+          # - alertmanager:9093
+
+# Load rules once and periodically evaluate them according to the global 'evaluation_interval'.
+rule_files:
+  - "*_rules.yml"
+  - "/sg_prometheus_add_ons/*_rules.yml"
+
+# A scrape configuration containing exactly one endpoint to scrape for metrics:
+scrape_configs:
+  # Scrape prometheus itself for metrics.
+  - job_name: "prometheus"
+    static_configs:
+        - targets: ["localhost:9090"]
+
+  - job_name: "sg"
+    file_sd_configs:
+      - files:
+          - "/sg_prometheus_add_ons/*_targets.yml"
+
+## Note: the content above is taken from sourcegraph/sourcegraph/docker-images/prometheus/config/prometheus.yml
+## I'm not currently sure how to configure the below in an external file via file_sd_configs.
+##
+## TODO(eric,uwedeportivo) - find a way to express what's below in an extensible way so we can 
+## just add config files in this directory instead of overwriting the entire main configuration.
```

Comment:
We can't merge this PR as-is with this in place. We must move towards less configuration in this repository and not more, introducing a new config file here is IMO not an option.

A few options I would be content with:

1. Merge this PR without monitoring for the workers, and defer this work to another PR. I am fine with that, but we would need to figure it out before 3.15 releases.
2. An environment variable on the prometheus image which controls whether or not this configuration is loaded (not a file stored in this repo).

Reply:
We can remove monitoring, but I'm not 100% we can fix this before the 3.15 release in the way @sourcegraph/code-intel wants (we're trying to rewrite parts of the worker in Go so we don't have this problem which is caused by running multiple node processes in the same container).

cc @uwedeportivo

Reply:
I don't think it would be too hard to solve, though?

Right now in `master` we are using the builtin `prometheus.yml` here: https://github.com/sourcegraph/sourcegraph/blob/master/docker-images/prometheus/config/prometheus.yml

You would just need to copy this prometheus.yml into the same location, but name it say e.g. `prometheus_pure_docker.yml` and then use it by setting the env var here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@6fef3d27e4938dd8d750afd7a81dd6a150d04e9b/-/blob/docker-images/prometheus/entry.sh#L4-8

That wouldn't be perfect (because you would still have `precise-code-intel-worker:9090` hard-coded in that file) but that'd be good enough for 3.15.

Then in e.g. `deploy-prometheus.sh` in this repo you would just add `-e PURE_DOCKER=true` to trigger using that config file.

Reply:
Ok that seems like a fine stopgap.

Reply:
Working on this now. And for context, this entire federation technique should be able to be removed in a few cycles once we rewrite some of the worker wrapper in Go and can use the language concurrency rather than the OS's.

Reply:
See https://github.com/sourcegraph/sourcegraph/pull/9552.

I'm unsure of how to update the prometheus image, though.